Add Mac App

conath opened the pull request:
I ported the main features to Mac! 😃 
You can see a list of trailers in a familiar layout and play them. This is an initial implementation without much attention given to the UI.

<img width="1920" alt="Screen Shot 2021-12-24 at 15 05 28" src="https://user-images.githubusercontent.com/12073163/147357929-48c11b59-297f-44be-abb0-2aec9427e770.png">

I am looking for feedback. This is my first Mac app! Edit: first one I made from scratch and using SwiftUI app lifecycle, anyway.

conath committed:
Add Mac target and basic working Mac app

The app only displays a list of movie titles.
The backend is shared with iOS, so some conditional compilation had to be added to the model.

conath committed:
Add JPEG compression in macOS model

conath committed:
Copy movie poster placeholder image to Mac target

conath committed:
Add FrontRow inspired trailer list view

Mac ContentView now shows image for selected movie
Includes a barely functional selection UI that needs to be adjusted to support scrolling

conath committed:
Update Mac App Icon

Make the film strip perforation and frames equally spaced vertically from the top and bottom of the icon

conath committed:
Fix Sound effects cause playing media to be paused #47

Set default AudioSession category to ambient (means it doesn't pause playing system-wide media)
Update AudioSession category when trailer start and stops playing

conath committed:
Move AudioFeedback into new file to share code with Mac app

conath committed:
Add MovieTrailerListView with audio feedback and Keyboard controls

conath committed:
Fix AudioFeedback Mac build error

conath committed:
Fix initial selection state and sorting

conath committed:
Fix iOS 15 Cover Flow scrolling bug #48

conath committed:
Add fade in and out on Mac app launch/quit

This involved re-writing the key events code.

conath committed:
Add deinit to KeyEventHandling to prevent possible memory leak

conath committed:
Add Esc key handling to go back / quit

conath committed:
Fix Selection Indicator vertical offset changing when in fullscreen

conath committed:
Fix Selection Indicator offset while scrolling

conath committed:
Add Enter and numpad enter to play trailer

conath committed:
Hide Mac app window title bar and add bottom list spacing

conath committed:
Add MoviePosterView to Mac app

conath committed:
Make Mac app launch in full screen

conath committed:
Rewrite Mac Movie Poster fade in/out

Change MoviePosterView to accept Binding instead of State to fix image not updated
Add FadeInOutPosterView to remove fade logic from ContentView

conath committed:
Add Audio Feedback to new key actions

Esc and Quit trigger the "exit" sound;
playing a trailer (return) triggers "selection"

conath committed:
Remove unused TrailerPlayerView file from iOS app

conath committed:
Add Mac TrailerPlayerView

conath committed:
Remove obsolete image fading code

conath committed:
Implement Trailer Playing feature

Rewrite TrailerPlayerView to remove unnecessary code
Add TrailerPlayerView to ContentView as Overlay

conath committed:
Update project metadata for initial release

Remove file select entitlement
Enable Hardened Runtime
Enable Code Signing

conath committed:
Merge branch 'main' into dev-mac

conath committed:
Fix MoviePosterView file shared between Mac and iOS, fixes iOS build

Xcode did not copy the file when I option-dragged it to the Mac folder...
this fixes that by copying the current state of MoviePosterView to the Mac app,
then reverting the changes made to the iOS file.

conath committed:
Fix crash if poster images unavailable

winsmith commented on TMT Mac/Views/MoviePosterView.swift:
```diff
@@ -0,0 +1,76 @@
+        } else {
+            image = NSImage(named: "moviePosterPlaceholder")
+        }
+        var fadeImage = image.cgImage(forProposedRect: nil, context: nil, hints: nil)!
```
I managed to crash at this line multiple times with my wonky internet connection. I suggest you solve this differently: For example, you could have just one instance of MoviePosterView instead of multiple ones with different parameters, encapsulated in FadeInOutPosterView. It should consist of a `ZStack` with a placeholder. Then put the actuall image on top if it is not nil. You can make it fade in with the `transition`

conath replied:
Not sure how this line relates to the internet connection. Either we have a downloaded image, or we get the processed cgImage.
Either way this code is now removed because I switched to `SwiftUI.Image`.

winsmith commented on TMT Mac/Views/MoviePosterView.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+    
+    var body: some View {
+        var image: NSImage!
```
Not sure why you are using `NSImage` instead of `Image` here, I assume it is because of legacy reasons. If not, I suggest you check out `Image` :)

conath replied:
I use the NSImage class because I need to mask the image with the gradient image at runtime. I had initially attempted to use SwiftUI's `.mask` modifier to achieve this, but it did not work.
https://github.com/conath/TheatricalMovieTrailers/blob/f38864db0da5b219ec5aa024125a33dc58e1b57d/TMT%20Mac/Views/MoviePosterView.swift#L34

I just attempted to do this again and it didn't work. I changed to a transparent-to-white gradient image mask and applied it to the reflection image. It seems like transparent images are not treated as having a less than 1 alpha value in SwiftUI, because this code did not work:

```swift
        return GeometryReader { geo in
            HStack {
                Spacer()
                VStack(alignment: .center, spacing: reflectionDistance) {
                    Image(nsImage: image)
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .frame(maxHeight: geo.size.height * 2 / 3)
                    Image(nsImage: image)
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .mask(Image("posterMirrorImageGradientAlphaMask"))
                        .opacity(0.7)
                        .frame(maxHeight: geo.size.height * 2 / 3)
                        .rotationEffect(.degrees(180), anchor: .center)
                        .rotation3DEffect(
                            .degrees(180),
                            axis: (x: 0, y: 1, z: 0)
                        )
                        .blur(radius: blurReflection ? 1 : 0)
    // ...
```

However I did get it working by using the original image (gradient black to white) and applying the `.luminanceToAlpha()` modifier to the mask view:

```swift
    .mask(
        Image("posterMirrorImageGradientMask")
            .resizable()
            .aspectRatio(contentMode: .fit)
            .frame(maxHeight: geo.size.height / 2) // placeholder for testing
            .luminanceToAlpha()
    )
```
🏆

winsmith commented on TMT Mac/ContentView.swift:
```diff
@@ -0,0 +1,126 @@
+                FadeInOutPosterView(posterImage: $dataStore.posterImage)
+                .frame(width: 0.5*frame.size.width)
+                .frame(maxHeight: .infinity)
+                .offset(x: 0, y: frame.size.height * 0.15)
```
Call me a nitpicker, but I think these should be indented. I can recommend the excellent SwiftFormat if you want this done automatically.

conath replied:
Couldn't get that installed, but I used the re-indent feature in Xcode to fix it.

<img width="628" alt="Screen Shot 2022-01-05 at 16 04 47" src="https://user-images.githubusercontent.com/12073163/148239836-883d914a-35f0-4628-ada5-9b0092bdf87a.png">

winsmith commented on TMT Mac/ContentView.swift:
```diff
@@ -0,0 +1,126 @@
+struct ContentView: View {
+    @ObservedObject private var settings = Settings.instance
+    @EnvironmentObject private var dataStore: MovieInfoDataStore
+    @State var sortingMode = SortingMode.ReleaseAscending
```
Any reason this is the only `@State` var not private?

conath replied:
That's an accident.

winsmith commented on TMT Mac/Settings.swift:
```diff
@@ -0,0 +1,89 @@
+    static let playUISounds = "playUISounds"
+}
+
+class Settings: ObservableObject {
```
I love the idea of abstracting settings into its own object! Have you considered using the `AppStorage` binding to automatically save these values and remove some code?

conath replied:
TIL [AppStorage](https://www.hackingwithswift.com/quick-start/swiftui/what-is-the-appstorage-property-wrapper) exists - thanks!

conath replied:
I've opened #51 to remind myself about this, as I do not need to touch the Settings code right now.

winsmith commented on TMT Mac/Settings.swift:
```diff
@@ -0,0 +1,89 @@
+    static let playUISounds = "playUISounds"
+}
+
+class Settings: ObservableObject {
```
I love the idea of abstracting settings into its own object! Have you considered using the `AppStorage` binding to automatically save these values and remove some code?

winsmith commented on TMT Mac/Utilities/KeyEventHandling.swift:
```diff
@@ -0,0 +1,105 @@
+import Carbon.HIToolbox
+import SwiftUI
+
+struct KeyEventHandling: NSViewRepresentable {
```
This is really cool and I didn't know it existed :)

winsmith commented on TMT Mac/Views/MoviePosterView.swift:
```diff
@@ -0,0 +1,76 @@
+struct MoviePosterView: View {
+    private let filmPosterAspectRatio = CGFloat(0.7063020214)
+    
+    @State private var reflectionDistance: CGFloat
```
I feel like this doesn't have to be a `@State` and could just be a `var`, because it doesn't change over the course of the lifetime of the view, but I might be incorrect.

conath replied:
You are right!

winsmith commented on TheatricalMovieTrailers/Model/MovieInfoDataStore.swift:
```diff
@@ -7,10 +7,22 @@
 import TelemetryClient
 
 class MovieInfoDataStore: ObservableObject {
+    #if os(iOS)
+    typealias Image = UIKit.UIImage
+    let appGroupID = "group.cafe.chrisp.tmt"
+    #elseif os(macOS)
+    typealias Image = AppKit.NSImage
+    let appGroupID = "cafe.chrisp.tmt-group"
+    #endif
```
SwiftUI already gives you an `Image` abstraction. I'd strongly recommend using that instead of `UIImage`/`NSImage`.

conath replied:
This is a great tip. I had gone down the "path of least resistance" when writing the Mac app and just updated the existing code to use `NSImage` where needed. Now I've re-written it to use `SwiftUI.Image` where applicable and it's much cleaner.

winsmith reviewed (approved):
This is good code. The main thing that might bring you pain in the future is that a lot of code is very platform-specific where it doesn't need to be. Consider leaning more heavily into the abstractions that SwiftUI provided for you – e.g. the provided `Image` class instead of your own type alias, or the `AppStorage` binding.

Another example is the heavy use of GeometryReader. For some things this is absolutely necessary, but looking through the project it seems a yellow flag that you use this *so* much. Maybe leave it like this for now but consider relying less on it when you next rewrite a view.

One thing that I noticed is that you duplicated all files in order to make a Mac version of the app. I think this is unnecessary. Swift files can belong to multiple targets, and if they are literally the same code, they shouldn't be duplicated. My suggestion is that you have one top level group for *Shared* code, one for *Mac* code and one for *iOS* code. All code should start out in the *Shared* folder, and only when you need to have different code for iOS and Mac should you duplicate the file, make sure the new files only belong to one target each, and work on them separately.

Formatting: This is very nitpicky, but I'm a fan of uniform code formatting. Doing these reviews it helps me when code is always formatted in the same way. If you think so too, consider checking out SwiftFormat, either the Xcode plugin or the command line app. 

I also pointed out some smaller individual changes where I saw them. All in all this is a very cool project, thank you for letting me have a loot :)

conath committed:
Update MoviePosterView to use SwiftUI Image for the reflection mask

Replace usage of NSImage with SwiftUI Image in Mac Views
(this likely breaks iOS build, will fix later)
Update Movie Poster Mirror Gradient Mask image
Fix ContentView variables not private

conath committed:
Fix code indentation in Mac files

conath commented:
Thank you very much @winsmith for the kind and helpful feedback. I have learned some new techniques from this and will apply them in the future. For example, the reliance on GeometryReader is definitely an anti-pattern I needed to be made aware of.

On the code re-use topic, I will keep it simple and finish the Mac App first before I look at ways to merge the view code. I have a rule of thumb for my code where I will allow one duplicate instance (so two instances of similar or the same code) and only refactor when a third instance would be created. This avoids unnecessary refactoring.

That being said, with the changes I made to remove the NSImage/UIImage code from `MoviePosterView`, I may be able to share that view with the iOS app now. I like the idea of an Xcode group with the shared code.

winsmith commented:
Awesome, glad I could help :D~

conath committed:
Update TelemetryClient package to 1.1.6

conath committed:
Fix Mac App Metadata for TestFlight upload

The macOS group ID must start with the Team ID, apparently.

conath committed:
Fix no fade in if trailers not loaded

conath committed:
Add non-exempt-encryption declaration to Mac Info.plist

conath committed:
Move code shared between platforms into Shared folder

conath committed:
Add mouse hover interaction, Fix mouse click interaction

Movie Trailer is selected when hovering over a row with the mouse cursor
Mouse cursor hides when using keyboard to navigate, this is to prevent hover selection by the mouse cursor just sitting there

conath committed:
Fix window focus by making SwiftUI Window key on launch

The issue was that key input and mouse hover was not working until you clicked into the fullscreen window.
I am not sure if a Mac restart or this code fixed it, but I am adding the code regardless.

conath committed:
Change version to 1.0-5

conath committed:
Move more of the shared files into Shared group

conath committed:
Fix iOS Build by replacing usages of UIImage with Image

Some other changes were needed along the way.

conath committed:
Update bundled trailers XML file

conath committed:
Add MovieInfoContainerView and simplify view structure

conath committed:
Add Movie poster scale animation

conath committed:
Implement showing basic Movie Metadata

conath committed:
Add true black background to the app on Mac

conath committed:
Use custom Font on Mac and show more Metadata

conath committed:
Disallow List View hover and some keyboard actions while trailer is playing

conath committed:
Add Screenshot of Mac App (Beta)

winsmith reviewed (approved):
I like this. It's well written and understandable.

This is not the scope of this PR, but in the future, consider using a [MainActor](https://www.avanderlee.com/swift/mainactor-dispatch-main-thread/) for your data store :) It'll make your code async ready and future proof

conath commented:
Thanks! I didn’t realize you were still getting notified about this. I will definitely look into async await and related topics at some point, as it’s come up multiple times in different projects. I guess it’s about time to merge this into the main branch and push for an App Store release at some point.

Unrelated, but I might rebrand the whole thing to Front Row Trailers, do you have any thoughts on that? Apple‘s trademark expired some years ago so it should be fine.

conath reviewed (commented):
Note to self: this code is wrong. If dark mode is not set, it still adds a black background. 

Edit: fixed
